---
title: "txseq: Exploratory data analysis report"
header-includes: 
- \usepackage{float} 
- \floatplacement{figure}{H} 
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: spacelab
    highlight: tango
    fig_caption: true
    code_folding: hide
urlcolor: blue
---

```{r exp-setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(out.extra = '')
```


```{r message=FALSE}
library(ggplot2)
library(circlize)
library(ComplexHeatmap)
library(gridExtra)
library(DESeq2)
library(reshape2)
library(dplyr)
library(vsn)
library(gplots)
library(RColorBrewer)
library(ggrastr)
library(txseq)
library(umap)
```


# Report configuration

```{r}
# Report configuration
setwd("~/work/txseq/hsc_example/reports/")

# Set the location of the txseq project run
txseq_run = "~/work/txseq/hsc_example"

# If you have used non-standard folder names you may need
# to also edit these paths.
sample_table = file.path(txseq_run,"samples.tsv")
bamqc_db = file.path(txseq_run, "bamqc/csvdb")
salmon_db =  file.path(txseq_run, "salmon/csvdb")

# configure the plots
# each variable should be set to either
# (a) a value recognised by ggplot OR
# (b) the name of a variable present in the samples.tsv file
plot_color="cell_type"    #e.g. "timepoint"
plot_shape="group"    #e.g. "knockdown"
plot_label="replicate"    #e.g."replicate"

exclude <- c()

# The number of replicates to require genes to be detected in
nreplicates=2
```


```{r include=FALSE, echo=FALSE}
# Load in the tximeta file containing the count data from Salmon
edata <- readRDS(file.path(txseq_run,"salmon/tximeta.dir/tximeta.RDS"))

gse <- edata$gse

# ** The workflow shown here assumes that we have counts from fragments of full length RNA
#    and not e.g. 3' lexogen data. **
#
# From the DESeq2 vignette:
#   If you have 3â€™ tagged RNA-seq data, then correcting the counts for gene length will induce a bias in your analysis, 
#   because the counts do not have length bias. Instead of using the 
#   default full-transcript-length pipeline, we recommend to use the original counts, e.g. txi$counts as a counts matrix, 
#   e.g. providing to DESeqDataSetFromMatrix or to the edgeR or # limma functions without calculating an offset and 
#   without using countsFromAbundance.

# set a grouping variable which combines cell type and condition

# set a grouping variable which combines cell type and condition
colData(gse)$group <- factor(paste(colData(gse)$cell_type,colData(gse)$condition,sep="_"))

dds <- DESeqDataSet(gse,
                    design=formula("~group"))

dds <- estimateSizeFactors(dds) 
vstMat <- assay(varianceStabilizingTransformation(dds, blind=TRUE, fitType="local"))
lq <- quantile(vstMat,0.25)
filteredVSTMat <- vstMat[apply(vstMat,1,function(x) sum(x[x>lq])>=nreplicates),]

# compute the PCA
pca <- prcomp(t(filteredVSTMat))


```


Exploratory analyses of the differences between the samples. The [DESeq2](https://www.bioconductor.org/packages/devel/bioc/html/DESeq2.html) package is used to prepare normalised, variance-stabilised gene-expression values (from read counts generated with [Salmon](https://combine-lab.github.io/salmon/) for downstream analysis.


# Number of counts and genes detected per-sample

```{r counts-vs-genes, fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}

# get per sample counts
d <- as.data.frame((apply(assays(gse)$counts,2,sum)))
colnames(d) <- "counts"

# get no. genes detected
d$no_genes <- (apply(assays(gse)$counts,2,function(x) length(x[x>=1])))

d$id <- factor(rownames(d), levels=rownames(d)[order(d$counts)])
d$group <- colData(gse)[rownames(d),plot_color]
d$label <- colData(gse)[rownames(d),plot_label]
d$shape <- colData(gse)[rownames(d),plot_shape]

gp <- ggplot(d,aes_string("counts","no_genes",color="group", shape="shape", label="label"))
gp <- gp + geom_text(nudge_y=160,color="grey")
gp <- gp + geom_point(size=4) + xlab("total read counts") + ylab("no. genes expressed")
gp <- gp + theme_bw()
print(gp)

```

# Data normalisation

The figure shows the result of applying variance stablizing transformation. Top panels: histograms of raw counts, middle panels: histograms of vst-transformed counts, bottom panel: relationship of standard deviation and expression level in vst-transformed data.

```{r exp-assess-norm,  fig.height=6, fig.width=8,  echo=FALSE, message=FALSE}

# assess the VST transformed normalised data:
counts <- counts(dds)
colnames(counts) <- dds$sample_id

c_gp <- plotDensities(counts, colData(dds), "group", "raw counts (log2)")
vst_gp <- plotDensities(2^vstMat, colData(dds), "group", "VST transformed, normalized counts")

notAllZero <- (rowMeans(vstMat)> min(vstMat))
sd_gp <- meanSdPlot(vstMat[notAllZero,], ylab="sd, vst",plot=F)

gps <- list(c_gp$den,c_gp$bp, vst_gp$den, vst_gp$bp, rasterise(sd_gp$gg + theme_bw()))

grid.arrange(grobs=gps, ncols=2, nrows=3)
```


```{r echo=FALSE, message=FALSE,include=FALSE}

# filter the vstMatrix
lq <- quantile(vstMat,0.25)
filteredVSTMat <- vstMat[apply(vstMat,1,function(x) sum(x[x>lq])>=nreplicates),]
print(dim(filteredVSTMat))
```

# Hierarchical clustering of samples

## Hierarchical clustering by correlation distance

The symmetrical heatmap shows the inter-sample pearson correlation co-efficients. The correlations were calculated using genes detected (above the 25^th^ expression quantile) in at least `r toString(nreplicates)` samples (n=`r toString(dim(filteredVSTMat)[1])`).

```{r exp-cor, fig.height=6, fig.width=6,  echo=FALSE, message=FALSE, eval=TRUE}

cor_method = "pearson"

m <- cor(filteredVSTMat, method=cor_method,  use="all")

distance_method <- paste(cor_method,"correlation")

condition = colData(dds)[colnames(m),"condition"]
celltype = colData(dds)[colnames(m),"cell_type"]

ha = rowAnnotation(condition=factor(condition),
                   celltype=factor(celltype),
                   col=list(condition=c("sscontrol"="lightgreen","curdlan"="red"),
                            celltype=c("lthsc"="darkgrey","sthsc"="lightgrey","mpp"="blue","gmp"="orange")))

xcol <- colorRamp2(c(quantile(m,0.1),quantile(m,0.5),quantile(m,0.9)),
                   c("blue","yellow","red"))

Heatmap(m,
        col=xcol,
        name="pearson\ncorrelation",
        right_annotation = ha,
        show_row_names = F,
        column_names_gp = gpar(fontsize = 6),
        show_column_names = T)


```


## Hierarchical clustering by manhattan distance

The symmetrical heatmap shows the inter-sample manhattan distances.

```{r exp-dist, fig.height=6, fig.width=6,  echo=FALSE, message=FALSE, eval=TRUE}

dm <- dist(t(filteredVSTMat), method="manhattan", diag=T, upper=T)
dm <- as.matrix(dm)

xcol <- colorRamp2(c(quantile(dm,0.1),quantile(dm,0.5),quantile(dm,0.9)),
                   c("blue","yellow","red"))

Heatmap(dm,
        col=xcol,
        name="manhattan\ndistance",
        right_annotation = ha,
        show_row_names = F,
        column_names_gp = gpar(fontsize = 6),
        show_column_names = T)

```


# Principle components analysis

## PCA of samples based on expressed genes

PCA analysis of samples based on genes detected (above the 25^th^ expression quantile) in at least `r toString(nreplicates)` samples (n=`r toString(dim(filteredVSTMat)[1])`).


```{r exp-pca-exprs, fig.height=6, fig.width=10,  echo=FALSE, message=FALSE}


gps <- ggplot_prcomp(pca, 
             plots=list("A"=c("PC1","PC2"), 
                        "B"=c("PC3","PC4"), 
                        "C"=c("PC5","PC6")),
             colData(gse), 
             color=plot_color, 
             shape=plot_shape,
             label=plot_label,
             size=2,
             nudge_scale_factor=20)

grid.arrange(grobs=gps,ncol=2)

```


## PCA based on genes with signicant variation between the experimental groups

```{r echo=FALSE, message=FALSE, include=FALSE}
# get genes showing significant variation
dds = DESeq(dds, test = "LRT", reduced = ~ 1)
res <- data.frame(results(dds))

print(dim(res))
sigvar_genes <- rownames(res)[res$padj<0.1 & !is.na(res$padj)]
sigVst <- vstMat[sigvar_genes,]
dim(sigVst)

```


PCA analysis of samples based on `r toString(dim(sigVst)[1])` genes that show significant variation between the experimental conditions (DESeq, LRT, 10% FDR).


```{r exp-pca-sig, fig.height=6, fig.width=10,  echo=FALSE, message=FALSE}
pca <- prcomp(t(sigVst))

gps <- ggplot_prcomp(pca, 
             plots=list("A"=c("PC1","PC2"), 
                        "B"=c("PC3","PC4"), 
                        "C"=c("PC5","PC6")),
             colData(gse), 
             color=plot_color, 
             shape=plot_shape,
             label=plot_label,
             size=2,
             nudge_scale_factor=20)

grid.arrange(grobs=gps,ncol=2)

```


# UMAP projection of samples based genes with significantly variable expression between the experimental groups

The projection is based on the expression of genes found to have significantly variable expression between the experimental groups (see above).

```{r echo=FALSE, message=FALSE}

umap_out = umap(t(filteredVSTMat))#, pca=T, perplexity=10)
```


```{r exp-tsne, fig.height=4, fig.width=6,  echo=FALSE, message=FALSE}

umap_frame <- as.data.frame(umap_out$layout)
rownames(umap_frame) <- colnames(filteredVSTMat)
colnames(umap_frame) <- c("UMAP1", "UMAP2")

at <- merge(umap_frame, colData(gse), by=0)

gp <- ggplot(at, aes_string("UMAP1", "UMAP2", 
                            color=plot_color, shape=plot_shape))
gp <- gp + geom_text(aes_string(label=plot_label), nudge_x=.1, nudge_y=.1,color="grey")
gp <- gp + geom_point(size=2,alpha=1)
gp <- gp + theme_bw()

print(gp)
```


# Hierarchical clustering of genes and samples


```{r echo=FALSE, message=FALSE, include=FALSE}
# retrieve the salmon_tpms
statement = 'select distinct s.* 
               from salmon_genes_tpms s 
               inner join transcript_info i
               on s.gene_id=i.gene_id
               where i.gene_biotype="protein_coding"'

raw_tpms = fetch_DataFrame(statement,salmon_db)

rownames(raw_tpms) <- raw_tpms$gene_id
raw_tpms$gene_id <- NULL

# perform upper quartile normalisation
tpms <- upperQuartileNormalise(raw_tpms)

```

```{R echo=FALSE, include=FALSE}
print(dim(tpms))
log_tpms <- log2(tpms + 1)
lq <- quantile(log_tpms,0.25)
filt_tpms <- log_tpms[apply(log_tpms,1,function(x) sum(x[x>lq])>nreplicates),]
dim(filt_tpms)
```


# Hierarchical clustering of samples based on protein-coding gene expression

Hierarchical clustering of genes and samples based on `r toString(nrow(filt_tpms))` expressed (>25^th^ expression quantile in at least `r toString(nreplicates)` samples) protein coding genes.

```{r exp-hclust-expr, fig.height=8, fig.width=8,  echo=FALSE, message=FALSE}

m <- filt_tpms

colden <- get_den(t(m), dist_method="manhattan", clust_method="complete", optimize=T)
rowden <- get_den(m, dist_method="manhattan", clust_method="complete", optimize=F)

hm <- function() {
    rnaseq_heatmap(m,
              labRow=F,
              colden=colden,
              rowden=rowden,
              key.xlab = "tpms log2(n+1)",
              key.xticks=c(1:16),
              log_factor=2)
    }


hm()
```


# Hierarchical clustering of samples based on significantly variable genes

Hierarchical clustering of genes and samples based on the expression of genes with significant variation in expression between the experimental groups.

```{r exp-hclust-sig, fig.height=8, fig.width=8,  echo=FALSE, message=FALSE}

m <- log_tpms[rownames(log_tpms) %in% rownames(sigVst),]

colden <- get_den(t(m), dist_method="manhattan", clust_method="complete", optimize=T)
rowden <- get_den(m, dist_method="manhattan", clust_method="complete", optimize=F)

hm <- function() {
    rnaseq_heatmap(m,
              labRow=F,
              colden=colden,
              rowden=rowden,
              key.xlab = "tpms log2(n+1)",
              key.xticks=c(1:16),
              log_factor=2)
    }


hm()
```

# Session information

```{r}
sessionInfo()
```



