---
title: "FASTQC summary report"
header-includes: 
- \usepackage{float} 
- \floatplacement{figure}{H} 
output:
  pdf_document:
    fig_caption: true
    keep_tex: true
    latex_engine: xelatex
  html_document:
    df_print: paged
urlcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(out.extra = '')
```

# FASTQC Summary report {#fastqc}

This report contains an aggregated overview of the QC metrics generated by the [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) package for each fastq file.

```{r include=FALSE, echo=FALSE}
library(ggplot2)
library(gridExtra)
library(dplyr)
library(txseq)
library(reshape2)
```


## Report configuration

```{r}
# location of the pipeline_fastqc.py sqlite database
fastqc_sqlite_database="/well/sansom/users/odq545/work/txseq/hsc_example/fastqc/csvdb"

# configure the plots
# each variable should be set to either
# (a) a value recognised by ggplot OR
# (b) the name of a variable present in the samples.tsv file
plot_color="cell_type"    #e.g. "timepoint"
plot_shape="condition"    #e.g. "knockdown"


# Seperate plots should be made for each end and lane.
# e.g. for paired-end sequenced on a single lane use: wrap_formula = "~end"
# e.g. for paired-end data sequenced on multiple lanes use: wrap_formula = "~end+lane"
# e.g. for single-end data sequenced on one lanes, set this variable to NULL to switch off wrapping: wrap_formula = NULL
# e.g. for single-end data sequenced on multiple lanes use: wrap_formula = "~lane"
wrap_formula = "~end"

```


## Core plots.

The data for the plots in this section should always be generated by the pipeline.

### Sequence quality histograms

A [Phred](https://en.wikipedia.org/wiki/Phred_quality_score) score of ~40 is expected for good quality bases. Base qualities may tail off toward the end of the read and be worse for the second read.


```{r seq-quality, fig.cap='sequence quality histograms', fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}

pbsq <- fetchFastQC(qc_metric="per_base_sequence_quality")

if(!is.null(pbsq))
{

gp <- ggplot(pbsq, aes_string("Base", "Mean", group="seq_id", color=plot_color,
                              shape=as.factor(plot_shape)))

gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula)) }
gp <- gp + xlab("Position in read") + ylab("Mean quality score")
gp <- gp + theme_bw()
print(gp)
}

```

\newpage


### Per-sequence Quality Scores

Ideally a peak around a [Phred](https://en.wikipedia.org/wiki/Phred_quality_score) score of 40 is expected.

```{r seq-scores, fig.cap='per sequence quality scores', fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}

psqs <- fetchFastQC(qc_metric="per_sequence_quality_scores")
if(!is.null(psqs))
{
gp <- ggplot(psqs, aes_string("Quality", "Count", 
                              group="seq_id", 
                              color=plot_color, 
                              shape=as.factor(plot_shape)))

gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + xlab("phred score") + ylab("count")
gp <- gp + theme_bw()
print(gp)
}
```

\newpage

### Per-base sequence content

Nucleotide frequency should be largely uniform across the reads.

```{r per-base, fig.cap='per base sequence content', fig.height=8, fig.width=8,  echo=FALSE, message=FALSE}
pbsc <- fetchFastQC(qc_metric="per_base_sequence_content")

plot_pbsc <- function(pbsc, nucleotide="A")
    {
gp <- ggplot(pbsc, aes_string("Base", nucleotide, 
                              group="seq_id", color=plot_color, 
                              shape=as.factor(plot_shape)))
gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free") }
gp <- gp + xlab("base position") + ylab(paste0("% ",nucleotide))
gp <- gp + ggtitle(nucleotide)
gp <- gp + theme_bw()
return(gp)}

if(!is.null(data))
{
gps <- list()
for(base in c("A","T","C","G")) { gps[[base]] <- plot_pbsc(pbsc,base) }

grid.arrange(grobs=gps,ncol=1)
}
```

\newpage

### Per-sequence GC content

An approximately normal distribution is expected.

```{r gc-content, fig.cap='per sequence GC content', fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}

psgc <- fetchFastQC(qc_metric="per_sequence_gc_content")

if(!is.null(psgc))
{
gp <- ggplot(psgc, aes_string("GC_Content", "Count", 
                              group="seq_id", color=plot_color, 
                              shape=as.factor(plot_shape)))
gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + xlab("mean GC content (%)") + ylab("count")
gp <- gp + theme_bw()
print(gp)
}
```


### Per-base N content

N's are assigned when a base's identity cannot be determined. Peaks may be seen at the beginning and end of reads.

```{r n-content, fig.cap='per base n content', fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}
pbnc <- fetchFastQC(qc_metric="per_base_n_content")

if(!is.null(pbnc))
{
gp <- ggplot(pbnc, aes_string("Base", "N_Count", 
                              group="seq_id", color=plot_color, 
                              shape=as.factor(plot_shape)))
gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + xlab("Position in read (bp)") + ylab("count")
gp <- gp + theme_bw()
print(gp)
}
```

\newpage

### Number of sequences

```{r seq-counts, fig.cap='sequence length distribution', fig.height=4, fig.width=8,  echo=FALSE, message=FALSE}

sld <- fetchFastQC(qc_metric="sequence_length_distribution")

if(!is.null(sld))
{
gp <- ggplot(sld, aes_string("seq_id", "Count", 
                             fill=plot_color, 
                             shape=as.factor(plot_shape)))
gp <- gp + geom_col(alpha=0.7) #+ geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + xlab("seq_id") + ylab("number of sequences")
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(gp)
}
```


### Sequence Length Distribution

If the reads have not been pre-processed (i.e. trimmed) a single peak representing the read-length is expected.

```{r seq-length, fig.cap='sequence length distribution', fig.height=3, fig.width=8,  echo=FALSE, message=FALSE}
if(!is.null(sld))
{
if(length(unique(sld$Length)) > 1)
{
sld$lenbin <- read.table(text=sld$Length,sep="-")$V1
} else { sld$lenbin <- sld$Length} 
 
gp <- ggplot(sld, aes_string("lenbin","Count", 
                             group="sample_id", color=plot_color, shape=plot_shape))
gp <- gp + geom_line(alpha=0.75) + geom_point()
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + ylab("number of sequences") + xlab("Sequence Length (bp)")
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(gp)
}
```

\newpage

### Sequence Duplication Levels

Ideally most of the sequences should be unique (i.e. having a duplication level of 1) - Figure ~\ref{fig:rad} shows the expected fraction of sequences remaining after deduplication.

Failure of the duplication level distribution to flatten when plotted as "percentage of deduplicated" indicates the present of many highly duplicated sequences.

```{r seq-duplication, fig.cap='sequence duplication levels', fig.height=6, fig.width=8,  echo=FALSE, message=FALSE}

sdl = fetchFastQC(qc_metric="sequence_duplication_levels")

if(!is.null(sdl))
{
sdl$Duplication_Level <- factor(as.character(sdl$Duplication_Level),
                                levels=c("0","1","2","3","4","5","6","7","8","9",
                                         ">10",">50",">100",">500",
                                         ">1k",">2k",">5k",">10k"))

plot_sdl <- function(sdl, sequences)
    {
gp <- ggplot(sdl, aes_string("Duplication_Level", sequences,
                             group="sample_id", 
                             color=plot_color, 
                             shape=plot_shape))
gp <- gp + geom_line(alpha=0.75) + geom_point(size=2, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + xlab("Sequence Duplication Level") + ylab("%")
gp <- gp + ggtitle(sequences)

gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    return(gp)
    }

gps <- list()
gps$pt <- plot_sdl(sdl, "Percentage_of_total")
gps$pd <- plot_sdl(sdl, "Percentage_of_deduplicated")


grid.arrange(grobs=gps,ncol=1)
}
```


### Percentage of reads remaining after de-duplication

```{r pct-nondup, fig.cap='percentage reads after de-duplication', fig.height=4, fig.width=8,  echo=FALSE, message=FALSE}

if(!is.null(sdl))
{
    
sdl2 <- sdl[sdl$Duplication_Level %in% c("1"),]

sdl2$Percent_after_deduplication <- sdl2$Percentage_of_total / sdl2$Percentage_of_deduplicated * 100
sdl2$sample <- factor(sdl2$sample, levels=sdl2$sample[rev(order(sdl2$Percent_after_deduplication))])

gp <- ggplot(sdl2, aes_string("seq_id", "Percent_after_deduplication",
                              group="seq_id", color=plot_color, 
                              shape=plot_shape))
gp <- gp + geom_point(size=4, alpha=0.75)
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }
gp <- gp + theme(axis.text.x = element_text(angle = 45, hjust = 1))

gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
print(gp)
}
```

\newpage

### Over represented sequences

Ideally there should be no or few over-represented sequences and these should represent only a small percentage of the reads. The sequence of over-represented reads is avaliable in the individual FastQC reports.

```{r ors-seq, fig.cap='over-represented sequences per sample', fig.height=4, fig.width=8,  echo=FALSE, message=FALSE}
ors <- fetchFastQC(qc_metric="overrepresented_sequences")

if(!is.null(ors))
{

x <- ors %>% group_by(seq_id, end, lane) %>% summarise(total_pct_over_rep_seq=sum(Percentage))

x$o <- paste(x$seq_id,x$end,sep="_")
x$o <- factor(x$o, levels=x$o[rev(order(x$total_pct_over_rep_seq))])

gp <- ggplot(x, aes_string("o","total_pct_over_rep_seq",
                              group="seq_id")) #, fill=plot_color, shape=plot_shape))
gp <- gp + geom_col()
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_x") }
gp <- gp + scale_x_discrete(breaks=x$o, labels=x$seq_id)
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
gp <- gp + ylab("total % of over-represented sequences") +xlab("seq_id")
gp


}
```


```{r ors-seq-by-contaminant, fig.cap='over-represented sequences by contaminant', fig.height=5, fig.width=8,  echo=FALSE, message=FALSE}
ors <- fetchFastQC(qc_metric="overrepresented_sequences")

if(!is.null(ors))
{

x <- ors %>% group_by(Possible_Source,end) %>% summarise(total_pct_over_rep_seq=mean(Percentage))

x$o <- paste(x$Possible_Source,x$end,sep="_")
x$o <- factor(x$o, levels=x$o[rev(order(x$total_pct_over_rep_seq))])

gp <- ggplot(x, aes_string("o","total_pct_over_rep_seq"))
gp <- gp + geom_col()
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_x") }
gp <- gp + scale_x_discrete(breaks=x$o, labels=x$Possible_Source)
gp <- gp + ylab("mean percentage of over-represented sequences") + xlab("possible source")
gp

}
```

\newpage

### Adapter content

Reads are examined for the presence of given adapter sequences.

```{r adapter-content, fig.cap='Adapter content', fig.height=5, fig.width=8,  echo=FALSE, message=FALSE}
ac = fetchFastQC(qc_metric="adapter_content")

if(!is.null(ac))
{


adapters <- c(colnames(ac)[grepl("_Adapter$",colnames(ac))],
              colnames(ac)[grepl("_Sequence$",colnames(ac))])

x <- melt(ac[,c("seq_id","end",adapters)], id.vars=c("seq_id","end"))
colnames(x) <- c("seq_id","end", "adaptor", "value") 


x$adaptor <- gsub("'","prime",x$adaptor)


x <- x %>% group_by(adaptor,end) %>% summarise(mean_adaptor_content=mean(value))
x$o <- paste(x$adaptor, x$end, sep="-")
x$o <- factor(x$o, levels=x$o[rev(order(x$mean_adaptor_content))])

#x$adaptor <- factor(x$adaptor, levels=x$adaptor[rev(order(x$mean_adaptor_content))])

gp <- ggplot(x, aes_string("o","mean_adaptor_content"))
gp <- gp + geom_col()
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_x") }
gp <- gp + scale_x_discrete(breaks=x$o, labels=x$adaptor)
gp <- gp + ylab("mean percentage of over-represented sequences") + xlab("adaptor")
print(gp)



}
```


## Additional plots

### Kmer content

To enable this plot, make sure the FastQC Kmer module is enabled in the limits file (the location of a custom limits file can be set in the pipeline yml file).

Presence of highly over-represented kmers may be, for example, caused by over-represented sequences, low-complexity libraries or a problem with the sequencing reaction.

```{r kmer-content, fig.cap='kmer content', fig.height=4, fig.width=8, echo=FALSE, message=FALSE, eval=FALSE}
kc = fetchFastQC(qc_metric="kmer_content")

if(!is.null(kc))
{

x <- melt(kc[,c("seq_id","end","Sequence","Count")], id.vars=c("seq_id","end","Sequence"))

x <- x %>% group_by(Sequence,end) %>% summarise(mean_kmer_count=mean(value)) %>% arrange(desc(mean_kmer_count)) %>% ungroup() %>% head(20)

x$o <- paste(x$Sequence, x$end, sep="-")
x$o <- factor(x$o, levels=x$o[rev(order(x$mean_kmer_count))])


gp <- ggplot(x, aes_string("o","mean_kmer_count"))
gp <- gp + geom_col()
gp <- gp + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1))
gp <- gp + scale_x_discrete(breaks=x$o, labels=x$Sequence)
gp <- gp + ylab("mean count per sample (top 20 kmers)") + xlab("adapter")
print(gp)
}
```



## Visualise the per tile sequence qualities

Note: not all sequence runs have tile information, so do not worry if this plot is not generated.

The plot shows sequence quality by position in the sequence lane. The color indicates the deviation from the average quality (red=bad, blue=good).


```{r tile-quality, fig.cap='per tile sequence qualities', fig.height=4, fig.width=6,  echo=FALSE, message=FALSE, eval=FALSE}

ptsq <- fetchFastQC(qc_metric="per_tile_sequence_quality")
if(!is.null(ptsq))
{

tmp <- ptsq %>% group_by(Base, Tile, end) %>% summarise(quality = mean(Mean))

tmp$Tile <- as.factor(tmp$Tile)
lvls <- levels(tmp$Tile)

gp <- ggplot(tmp, aes(Tile,Base)) 
gp <- gp + geom_raster(aes(fill=quality))

if(!is.null(wrap_formula)) { gp <- gp + facet_wrap(formula(wrap_formula), scales="free_y") }

gp <- gp + scale_fill_gradient2(low="darkred", mid="white", high="darkblue", midpoint=0)
gp <- gp + scale_x_discrete(breaks=lvls[seq(1,length(lvls),by=10)])
gp <- gp + theme_bw()
print(gp)
}
```
